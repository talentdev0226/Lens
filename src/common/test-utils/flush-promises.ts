/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { setImmediate } from "timers";

export const flushPromises = () => new Promise(setImmediate);
