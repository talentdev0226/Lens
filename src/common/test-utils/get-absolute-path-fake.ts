/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { GetAbsolutePath } from "../path/get-absolute-path.injectable";

export const getAbsolutePathFake: GetAbsolutePath = (...args) => {
  const maybeAbsolutePath = args.join("/");

  if (isAbsolutePath(maybeAbsolutePath)) {
    return maybeAbsolutePath;
  }

  return `/some-absolute-root-directory/${maybeAbsolutePath}`;
};

const isAbsolutePath = (path: string) => path.startsWith("/");
