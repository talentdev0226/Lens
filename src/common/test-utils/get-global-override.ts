/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Injectable } from "@ogre-tools/injectable";

export interface GlobalOverride {
  injectable: Injectable<any, any, any>;
  overridingInstantiate: any;
}

export const getGlobalOverride = <T extends Injectable<any, any, any>>(
  injectable: T,
  overridingInstantiate: T["instantiate"],
) => ({
    injectable,
    overridingInstantiate,
  });
