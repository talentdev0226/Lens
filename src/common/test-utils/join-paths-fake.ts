/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { JoinPaths } from "../path/join-paths.injectable";

export const joinPathsFake: JoinPaths = (...args) => args.join("/");
