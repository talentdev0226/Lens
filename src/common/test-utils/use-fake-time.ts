/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { act } from "@testing-library/react";

let usingFakeTime = false;

export const advanceFakeTime = (milliseconds: number) => {
  if (!usingFakeTime) {
    throw new Error("Tried to advance fake time but it was not enabled. Call useFakeTime() first.");
  }

  act(() => {
    jest.advanceTimersByTime(milliseconds);
  });
};

export const useFakeTime = (dateTime = "2015-10-21T07:28:00Z") => {
  usingFakeTime = true;

  jest.useFakeTimers();

  jest.setSystemTime(new Date(dateTime));
};
