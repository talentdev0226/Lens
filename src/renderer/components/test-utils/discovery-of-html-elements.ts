/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { RenderResult } from "@testing-library/react";

type DiscoverySourceTypes = RenderResult | Element;

export type QuerySingleElement = (
  attributeName: string,
  attributeValue?: string
) => { discovered: Element | null } & Discover;

export type GetSingleElement = (
  attributeName: string,
  attributeValue?: string
) => { discovered: Element } & Discover;

export type QueryAllElements = (attributeName: string) => {
  discovered: Element[];
  attributeValues: (string | null)[];
};

export interface Discover {
  querySingleElement: QuerySingleElement;
  queryAllElements: QueryAllElements;
  getSingleElement: GetSingleElement;
}

export const discoverFor = (getSource: () => DiscoverySourceTypes): Discover => ({
  querySingleElement: querySingleElement(getSource),
  queryAllElements: queryAllElements(getSource),
  getSingleElement: getSingleElement(getSource),
});

export const querySingleElement =
  (getSource: () => DiscoverySourceTypes): QuerySingleElement =>
    (attributeName, attributeValue) => {
      const source = getSource();

      const dataAttribute = `data-${attributeName}-test`;

      const selector = attributeValue
        ? `[${dataAttribute}="${attributeValue}"]`
        : `[${dataAttribute}]`;

      const discovered = getBaseElement(source).querySelector(selector);

      const nestedDiscover = discoverFor(() => {
        if (!discovered) {
          throw new Error("Tried to do nested discover using source that does not exist");
        }

        return discovered;
      });

      return {
        discovered,

        ...nestedDiscover,
      };
    };

export const queryAllElements =
  (getSource: () => DiscoverySourceTypes): QueryAllElements =>
    (attributeName) => {
      const source = getSource();

      const dataAttribute = `data-${attributeName}-test`;

      const results = [
        ...getBaseElement(source).querySelectorAll(`[${dataAttribute}]`),
      ];

      return {
        discovered: results,

        attributeValues: results.map((result) =>
          result.getAttribute(dataAttribute),
        ),
      };
    };

export const getSingleElement =
  (getSource: () => DiscoverySourceTypes): GetSingleElement =>
    (attributeName, attributeValue) => {
      const dataAttribute = `data-${attributeName}-test`;

      const { discovered, ...nestedDiscover } = querySingleElement(getSource)(
        attributeName,
        attributeValue,
      );

      if (!discovered) {
        const validValues =
        queryAllElements(getSource)(attributeName).attributeValues;

        if (attributeValue) {
          throw new Error(
            `Couldn't find HTML-element with attribute "${dataAttribute}" with value "${attributeValue}". Present values are:\n\n"${validValues.join(
              '",\n"',
            )}"`,
          );
        }

        throw new Error(
          `Couldn't find HTML-element with attribute "${dataAttribute}"`,
        );
      }

      return { discovered, ...nestedDiscover };
    };

const getBaseElement = (source: DiscoverySourceTypes) =>
  "baseElement" in source ? source.baseElement : source;
