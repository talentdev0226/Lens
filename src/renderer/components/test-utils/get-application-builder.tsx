/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed, IObservableArray, observable, runInAction } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import { fireEvent, RenderResult } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import directoryForLensLocalStorageInjectable from "../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import { Sidebar } from "../layout/sidebar";
import { getDisForUnitTesting } from "../../../test-utils/get-dis-for-unit-testing";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import applicationMenuItemsInjectable, { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;
  addExtensions: (...extensions: LensRendererExtension[]) => Promise<ApplicationBuilder>;
  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeSetups: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigation: {
      click: (id: string) => void;
    };
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  renderSidebar: () => React.ReactNode;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const { rendererDi, mainDi, runSetups } = getDisForUnitTesting({
    doGeneralOverrides: true,
  });

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const extensionsState = observable.array<LensRendererExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments: Record<string, Environment> = {
    application: {
      renderSidebar: () => null,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    },

    clusterFrame: {
      renderSidebar: () => <Sidebar />,
      onAllowKubeResource: () => {},
    },
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => extensionsState),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed((): LensMainExtension[] => []),
  );

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click(undefined, undefined, undefined);
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join(
                '", "',
              )}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      rendererDi.override(
        directoryForLensLocalStorageInjectable,
        () => "/irrelevant",
      );

      return builder;
    },

    addExtensions: async (...extensions) => {
      const extensionRegistrators = rendererDi.injectMany(
        extensionRegistratorInjectionToken,
      );

      const addAndEnableExtensions = async () => {
        const registratorPromises = extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        );

        await Promise.all(registratorPromises);

        runInAction(() => {
          extensions.forEach((extension) => {
            extensionsState.push(extension);
          });
        });
      };

      if (rendered) {
        await addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }

      return builder;
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeSetups(callback: (dis: DiContainers) => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      for (const callback of beforeSetupsCallbacks) {
        await callback(dis);
      }

      await runSetups();

      const render = renderFor(rendererDi);

      const history = rendererDi.inject(observableHistoryInjectable);

      const currentRouteComponent = rendererDi.inject(
        currentRouteComponentInjectable,
      );

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      rendered = render(
        <Router history={history}>
          {environment.renderSidebar()}

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );

      return rendered;
    },
  };

  return builder;
};

const toFlatChildren =
  (parentId: string) =>
    ({
      submenu = [],
      ...menuItem
    }: MenuItemOpts): (MenuItemOpts & { path: string })[] =>
      [
        {
          ...menuItem,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        },
        ...submenu.flatMap(toFlatChildren(menuItem.id)),
      ];
