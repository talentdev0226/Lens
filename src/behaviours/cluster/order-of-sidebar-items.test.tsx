/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import { fireEvent, RenderResult } from "@testing-library/react";
import { SidebarItemRegistration, sidebarItemsInjectionToken } from "../../renderer/components/layout/sidebar-items.injectable";
import { computed } from "mobx";
import { get, includes, noop } from "lodash/fp";
import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";

describe("cluster - order of sidebar items", () => {
  let rendered: RenderResult;
  let applicationBuilder: ApplicationBuilder;

  beforeEach(() => {
    applicationBuilder = getApplicationBuilder().setEnvironmentToClusterFrame();

    applicationBuilder.beforeSetups(({ rendererDi }) => {
      rendererDi.register(testSidebarItemsInjectable);
    });
  });

  describe("when rendered", () => {
    beforeEach(async () => {
      rendered = await applicationBuilder.render();
    });

    it("renders", () => {
      expect(rendered.container).toMatchSnapshot();
    });

    it("has parent items in order", () => {
      const actual = rendered
        .queryAllByTestId("sidebar-item")

        .filter((element) =>
          includes(element.dataset.idTest)([
            "some-parent-id",
            "some-other-parent-id",
            "some-another-parent-id",
          ]),
        )

        .map(get("dataset.idTest"));

      expect(actual).toEqual([
        "some-parent-id",
        "some-another-parent-id",
        "some-other-parent-id",
      ]);
    });

    describe("when parent is expanded", () => {
      beforeEach(() => {
        const parentLink = rendered.getByTestId(
          "sidebar-item-link-for-some-parent-id",
        );

        fireEvent.click(parentLink);
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("has child items in order", () => {
        const actual = rendered
          .queryAllByTestId("sidebar-item")
          .filter(
            (element) => element.dataset.parentIdTest === "some-parent-id",
          )
          .map(get("dataset.idTest"));

        expect(actual).toEqual([
          "some-child-id",
          "some-another-child-id",
          "some-other-child-id",
        ]);
      });
    });
  });
});

const testSidebarItemsInjectable = getInjectable({
  id: "some-sidebar-item-injectable",

  instantiate: () =>
    computed((): SidebarItemRegistration[] => [
      {
        id: "some-parent-id",
        parentId: null,
        title: "Some parent",
        onClick: noop,
        orderNumber: 42,
      },
      {
        id: "some-other-parent-id",
        parentId: null,
        title: "Some other parent",
        onClick: noop,
        orderNumber: 126,
      },
      {
        id: "some-another-parent-id",
        parentId: null,
        title: "Some another parent",
        onClick: noop,
        orderNumber: 84,
      },
      {
        id: "some-child-id",
        parentId: "some-parent-id",
        title: "Some child",
        onClick: noop,
        orderNumber: 168,
      },
      {
        id: "some-other-child-id",
        parentId: "some-parent-id",
        title: "Some other child",
        onClick: noop,
        orderNumber: 252,
      },
      {
        id: "some-another-child-id",
        parentId: "some-parent-id",
        title: "Some another child",
        onClick: noop,
        orderNumber: 210,
      },
    ]),

  injectionToken: sidebarItemsInjectionToken,
});
