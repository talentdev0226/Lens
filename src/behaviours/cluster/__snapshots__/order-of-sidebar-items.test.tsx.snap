// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cluster - order of sidebar items when rendered renders 1`] = `
<div>
  <div
    class="flex flex-col"
    data-testid="cluster-sidebar"
  >
    <div
      class="SidebarCluster"
    >
      <div
        class="Avatar rounded loadingAvatar"
        style="width: 40px; height: 40px;"
      >
        ??
      </div>
      <div
        class="loadingClusterName"
      />
    </div>
    <div
      class="sidebarNav sidebar-active-status"
    >
      <div
        class="SidebarItem"
        data-id-test="workloads"
        data-is-active-test="false"
        data-test-id="workloads"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-workloads"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            />
          </i>
          <span
            class="link-text box grow"
          >
            Workloads
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="config"
        data-is-active-test="false"
        data-test-id="config"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-config"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="list"
            >
              list
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Config
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-parent-id"
        data-is-active-test="false"
        data-test-id="some-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-some-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some parent
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="network"
        data-is-active-test="false"
        data-test-id="network"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-network"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="device_hub"
            >
              device_hub
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Network
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="storage"
        data-is-active-test="false"
        data-test-id="storage"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-storage"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="storage"
            >
              storage
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Storage
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-another-parent-id"
        data-is-active-test="false"
        data-test-id="some-another-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-another-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some another parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="helm"
        data-is-active-test="false"
        data-test-id="helm"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-helm"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            />
          </i>
          <span
            class="link-text box grow"
          >
            Helm
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="user-management"
        data-is-active-test="false"
        data-test-id="user-management"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-user-management"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="security"
            >
              security
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Access Control
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="custom-resources"
        data-is-active-test="false"
        data-test-id="custom-resources"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-custom-resources"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="extension"
            >
              extension
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Custom Resources
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-other-parent-id"
        data-is-active-test="false"
        data-test-id="some-other-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-other-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some other parent
          </span>
        </a>
      </div>
    </div>
  </div>
  <div
    class="Notifications flex column align-flex-end"
  />
</div>
`;

exports[`cluster - order of sidebar items when rendered when parent is expanded renders 1`] = `
<div>
  <div
    class="flex flex-col"
    data-testid="cluster-sidebar"
  >
    <div
      class="SidebarCluster"
    >
      <div
        class="Avatar rounded loadingAvatar"
        style="width: 40px; height: 40px;"
      >
        ??
      </div>
      <div
        class="loadingClusterName"
      />
    </div>
    <div
      class="sidebarNav sidebar-active-status"
    >
      <div
        class="SidebarItem"
        data-id-test="workloads"
        data-is-active-test="false"
        data-test-id="workloads"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-workloads"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            />
          </i>
          <span
            class="link-text box grow"
          >
            Workloads
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="config"
        data-is-active-test="false"
        data-test-id="config"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-config"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="list"
            >
              list
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Config
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-parent-id"
        data-is-active-test="false"
        data-test-id="some-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-some-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some parent
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_up"
            >
              keyboard_arrow_up
            </span>
          </i>
        </a>
        <ul
          class="sub-menu"
        >
          <div
            class="SidebarItem"
            data-id-test="some-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-child-id"
            data-testid="sidebar-item"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                Some child
              </span>
            </a>
          </div>
          <div
            class="SidebarItem"
            data-id-test="some-another-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-another-child-id"
            data-testid="sidebar-item"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-another-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                Some another child
              </span>
            </a>
          </div>
          <div
            class="SidebarItem"
            data-id-test="some-other-child-id"
            data-is-active-test="false"
            data-parent-id-test="some-parent-id"
            data-test-id="some-other-child-id"
            data-testid="sidebar-item"
          >
            <a
              class="nav-item flex gaps align-center"
              data-testid="sidebar-item-link-for-some-other-child-id"
              href="/"
            >
              <span
                class="link-text box grow"
              >
                Some other child
              </span>
            </a>
          </div>
        </ul>
      </div>
      <div
        class="SidebarItem"
        data-id-test="network"
        data-is-active-test="false"
        data-test-id="network"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-network"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="device_hub"
            >
              device_hub
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Network
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="storage"
        data-is-active-test="false"
        data-test-id="storage"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-storage"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="storage"
            >
              storage
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Storage
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-another-parent-id"
        data-is-active-test="false"
        data-test-id="some-another-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-another-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some another parent
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="helm"
        data-is-active-test="false"
        data-test-id="helm"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-helm"
          href="/"
        >
          <i
            class="Icon svg focusable"
          >
            <span
              class="icon"
            />
          </i>
          <span
            class="link-text box grow"
          >
            Helm
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="user-management"
        data-is-active-test="false"
        data-test-id="user-management"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-user-management"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="security"
            >
              security
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Access Control
          </span>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="custom-resources"
        data-is-active-test="false"
        data-test-id="custom-resources"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center expandable"
          data-testid="sidebar-item-link-for-custom-resources"
          href="/"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="extension"
            >
              extension
            </span>
          </i>
          <span
            class="link-text box grow"
          >
            Custom Resources
          </span>
          <i
            class="Icon expand-icon box right material focusable"
          >
            <span
              class="icon"
              data-icon-name="keyboard_arrow_down"
            >
              keyboard_arrow_down
            </span>
          </i>
        </a>
      </div>
      <div
        class="SidebarItem"
        data-id-test="some-other-parent-id"
        data-is-active-test="false"
        data-test-id="some-other-parent-id"
        data-testid="sidebar-item"
      >
        <a
          class="nav-item flex gaps align-center"
          data-testid="sidebar-item-link-for-some-other-parent-id"
          href="/"
        >
          <span
            class="link-text box grow"
          >
            Some other parent
          </span>
        </a>
      </div>
    </div>
  </div>
  <div
    class="Notifications flex column align-flex-end"
  />
</div>
`;
